### DISCUSSION TITLE
Fastest way to search and filter for messages using IMAP

### DISCUSSION THREAD

**Discussion by lonix1** [jstedfast/MailKit#1862] [Q&A] [answered]
Using the library's IMAP functionality, I'd like to get a number of messages in the shortest time possible.

Suppose I'm searching/filtering by subject, e.g. `"Your case file (ref. af6177eb-83d1-4271-901b-e78c0a31cc1f)"`.

I'm using:
```cs
// search for messages using a query
var query = SearchQuery
  .DeliveredAfter(DateTime.UtcNow.AddDays(-1))
  .And(SearchQuery.SubjectContains("Your case file"))
  .And(SearchQuery.FromContains("no-reply@example.com"));
var soughtUniqueIds = await inbox.SearchAsync(query);

// fetch summaries for matching messages
var summaries = await inbox.FetchAsync(soughtUniqueIds, MessageSummaryItems.Envelope);

// further filter summaries using regex on subject
var filteredUniqueIds =
  summaries
  .Where(summary => IsRegexMatch(summary.Envelope.Subject))
  .Select(summary => summary.UniqueId);

// do something with filteredUniqueIds...
// process, print, delete, etc.
```

I compared two approaches:
- `inbox.FetchAsync(uniqueIds, MessageSummaryItems.Envelope)` and then using `summary.Envelope.Subject`, or
- `inbox.FetchAsync(uniqueIds, MessageSummaryItems.Headers, new HeaderId[1] { HeaderId.Subject })` and then using `summary.Headers.Single().Value`?

I benchmarked both and received similar but unpredictable results; probably server issues and latency.

Which is the "faster" approach, in theory?

Or perhaps there's an even faster way?

**Comment by jstedfast** [comment, marked as answer]
The ENVELOPE data structure is typically cached by the IMAP server (because it is both static data and designed to be commonly requested by clients), so it should theoretically be more likely to be faster than requesting headers which is going to require the server to parse each message.

Obviously it all comes down to the particular implementation of the IMAP server.

**Reply by lonix1** [reply]
Thanks. I'm not an IMAP expert, so I wasn't sure. Appreciate the advice, and this library.

Happy New Year!
